internal/graph/cycle_check_resolver.go

```diff
@@ -0,0 +1,55 @@
+package graph
+
+import (
+	"context"
+
+	"github.com/openfga/openfga/pkg/tuple"
+)
+
+type CycleDetectionCheckResolver struct {
+	delegate CheckResolver
+}
+
+var _ CheckResolver = (*CycleDetectionCheckResolver)(nil)
+
+// Close implements CheckResolver.
+func (*CycleDetectionCheckResolver) Close() {}
+
+func NewCycleDetectionCheckResolver() *CycleDetectionCheckResolver {
+	c := &CycleDetectionCheckResolver{}
+	c.delegate = c
+
+	return c
+}
+
+// ResolveCheck implements CheckResolver.
+func (c *CycleDetectionCheckResolver) ResolveCheck(
+	ctx context.Context,
+	req *ResolveCheckRequest,
+) (*ResolveCheckResponse, error) {
+	key := tuple.TupleKeyToString(req.GetTupleKey())
+
+	if req.VisitedPaths == nil {
+		req.VisitedPaths = map[string]struct{}{}
+	}
+
```

Comment:
we should do `maps.Clone(req.VisitedPaths)` here. It should not be the caller's responsibility to do this.